src/storage/src/source/source_reader_pipeline.rs

```diff
@@ -1277,287 +804,149 @@ where
-                            ts,
+                            into_ts,
                             id,
-                        )
+                        );
+                        total_processed += 1;
                     }
-                }
-
-                batch_capabilities.subtract(&untimestamped_batch.batch_lower);
-
-                // Pop off the processed batch.
-                untimestamped_batches.pop_front();
-            }
 
-            bytes_read_counter.inc_by(u64::cast_from(bytes_read));
-            source_metrics.record_partition_offsets(metric_updates);
-
-            // This is correct for totally ordered times because there can be at
-            // most one entry in the `CapabilitySet`. If this ever changes we
-            // need to rethink how we surface this in metrics. We will notice
-            // when that happens because the `expect()` will fail.
-            source_metrics.capability.set(
-                cap_set
-                    .iter()
-                    .at_most_one()
-                    .expect("there can be at most one element for totally ordered times")
-                    .map(|c| c.time())
-                    .cloned()
-                    .unwrap_or(Timestamp::MAX)
-                    .into(),
-            );
+                    let total_skipped = total_buffered - total_processed;
+                    trace!("timely-{worker_id} reclock({id}) processed {}, skipped {} messages", total_processed, total_skipped);
+
+                    bytes_read_counter.inc_by(u64::cast_from(bytes_read));
+                    source_metrics.record_partition_offsets(metric_updates);
+
+                    // This is correct for totally ordered times because there can be at
+                    // most one entry in the `CapabilitySet`. If this ever changes we
+                    // need to rethink how we surface this in metrics. We will notice
+                    // when that happens because the `expect()` will fail.
+                    source_metrics.capability.set(
+                        cap_set
+                            .iter()
+                            .at_most_one()
+                            .expect("there can be at most one element for totally ordered times")
+                            .map(|c| c.time())
+                            .cloned()
+                            .unwrap_or(mz_repr::Timestamp::MAX)
+                            .into(),
+                    );
 
-            // Update our view of the global source upper after we're done with
-            // everything else that needs processing.
-            let ingest_frontier = frontiers[0].frontier();
-            for (cap, (pid, offset)) in source_upper_updates.iter() {
-                if ingest_frontier.less_equal(cap) {
-                    continue;
-                }
+                    // We only want to retain non-empty batches
+                    untimestamped_batches.retain(|batch| !batch.is_empty());
```

Comment:
should this happen earlier? before we `handle_messages`? I assume its just here to reclock times with no messages?

Reply:
It can moved earlier but not before the loop that drains the iterator because that is the thing that will empty out batches. I moved it right after we're finished with the reclock loop